src/vos/vos_iterator.c

```diff
@@ -666,18 +657,21 @@ vos_iterate_internal(vos_iter_param_t *param, vos_iter_type_t type,
 	}
 	read_time = dtx_is_valid_handle(dth) ? dth->dth_epoch : 0 /* unused */;
 probe:
-	if (!daos_anchor_is_zero(anchor))
-		probe_anchor = anchor;
-
-	rc = vos_iter_probe(ih, probe_anchor);
+	rc = vos_iter_probe(ih, anchor);
+	if (rc == VOS_ITER_CB_YIELD) {
+		set_reprobe(type, VOS_ITER_CB_YIELD, anchors, 0);
+		D_ASSERT(need_reprobe(type, anchors));
```

Comment:
I suppose D_ASSERT() will never be fully optimized out by any compiler option, right? Cause we rely on the need_reprobe() call to clear the 'need_reprobe' flag on anchor.

Reply:
I don't think we plan to remove it

Reply:
> I suppose D_ASSERT() will never be fully optimized out by any compiler option, right? Cause we rely on the need_reprobe() call to clear the 'need_reprobe' flag on anchor.

Rename need_reprobe() to check_and_reset_reprobe() is more clear for me. Or split it into two functions.

Reply:
agreed, we should not call a function that does anything more than a boolean check within assertion.